qiskit_experiments/curve_analysis/curve_analysis.py

```diff
@@ -314,36 +317,29 @@ def _run_curve_fit(
         if isinstance(fit_options, FitOptions):
             fit_options = [fit_options]
 
-        valid_uncertainty = np.all(np.isfinite(curve_data.y_err))
-
-        model_weights = {}
-        if valid_uncertainty:
-            for model in models:
-                sub_yerr = curve_data.get_subset_of(model._name).y_err
-                if len(sub_yerr) == 0:
-                    continue
-                nonzero_yerr = np.where(np.isclose(sub_yerr, 0.0), np.finfo(float).eps, sub_yerr)
+        # Create convenient function to compute residual of the models.
+        partial_residuals = []
+        valid_uncertainty = np.all(np.isfinite(curve_data.yerr.to_numpy()))
+        for i, sub_data in list(curve_data.groupby("model_id")):
```

Comment:
Filtering for each subset is now implemented by dataframe, but this creates another dataframe for subset. Running this in the objective function which is called many times by the optimizer degrades performance. Instead, this logic filters subset in advance and create partial-bound functions that evaluate the residual. This alleviates the performance bottleneck.